### ISSUE TITLE
DALI TensorGPU's copy_to_external to pycuda slow

### ISSUE THREAD

**Issue by abcilike** [NVIDIA/DALI#1197]
Slow:
 device_mem = cuda.mem_alloc(size)
 c_type_pointer = ctypes.c_void_p(int(device_mem))
 time_start = time.time()
 dali_tensor.copy_to_external(c_type_pointer)
  print('copy 0 {}'.format(time.time() - time_start))
------------------------------
copy | 0.015529394149780273
copy | 0.016887187957763672
copy | 0.015774965286254883
copy | 0.015652894973754883

 Fast:
 arr = torch.zeros(category_shapes[category],
                                                          dtype=pyt_tensors[category].dtype,
                                                          device=pyt_tensors[category].device)
 c_type_pointer = ctypes.c_void_p(arr.data_ptr())
 time_start = time.time()
 dali_tensor.copy_to_external(c_type_pointer)
 print('copy | {}'.format(time.time() - time_start))
-----------------------------------------
copy | 0.0016980171203613281
copy | 0.0016565322875976562
copy | 0.00081634521484375
copy | 0.0008134841918945312

10~20x slower

**Comment by abcilike** [comment]
I don't know why, or did I make a mistake?

**Comment by abcilike** [comment]
copy data size is shape(416*416*3*64) np.float42, so gpu mem size is 416*416*3*64*4

**Comment by awolant** [comment]
Hi, thanks for the question.
I does look a bit off. At first glance I don't see any mistake in there, but it's hard to say. Could you provide a more complete reproduction, so it would be easier to investigate this issue?

**Comment by abcilike** [comment]
# image dir with plain jpeg files
image_dir = "/home/deep_cql/projects/data/pic_64/test2"
BATCH_SIZE = 64  # batch size per GPU
D_SIZE = 416

import os
os.environ["CUDA_DEVICE"] = "2"
import pycuda.driver as cuda
import pycuda.autoinit
import time
import nvidia.dali.ops as ops
import nvidia.dali.types as types
from nvidia.dali.pipeline import Pipeline
import torch
import ctypes

class CommonPipeline(Pipeline):
  def __init__(self, batch_size, num_threads, device_id):
    super(CommonPipeline, self).__init__(batch_size, num_threads, device_id, exec_async=True,
                                         exec_pipelined=True)

    self.input = ops.ExternalSource()
    self.input_label = ops.ExternalSource()
    self.decode = ops.ImageDecoder(device="mixed", output_type=types.RGB)
    self.resize = ops.Resize(device="gpu",
                             image_type=types.RGB,
                             interp_type=types.INTERP_LINEAR, resize_x=416, resize_y=416)
    self.cmn = ops.NormalizePermute(device="gpu",
                                    output_dtype=types.FLOAT,
                                    height=416,
                                    width=416,
                                    image_type=types.RGB,
                                    mean=[0., 0., 0.],
                                    std=[255., 255., 255.])

  def base_define_graph(self, inputs, labels):
    images = self.decode(inputs)
    # images = self.mix_and_split(images)
    images = self.resize(images)
    output = self.cmn(images)
    return (output, labels)

class FileReadPipeline(CommonPipeline):
  def __init__(self, batch_size, num_threads, device_id):
    super(FileReadPipeline, self).__init__(batch_size, num_threads, device_id)
    self.input = ops.FileReader(file_root=image_dir)
    # self.input = ops.FileReader(file_root=image_dir)

  def define_graph(self):
    images, labels = self.input(name="Reader")
    return self.base_define_graph(images, labels)

if __name__ == '__main__':
  pipe_t = [FileReadPipeline, (0, 1)]
  pipe_name, label_range = pipe_t
  print("RUN: " + pipe_name.__name__)

  pipes = pipe_name(batch_size=BATCH_SIZE, num_threads=3, device_id=2)
  pipes.build()
  pipe_out = pipes.run()
  print(pipe_out)
  dali_tensor, _ = pipe_out
  for i in range(20):
    # slow
    size = D_SIZE*D_SIZE*3*BATCH_SIZE*4
    device_mem = cuda.mem_alloc(size)
    c_type_pointer = ctypes.c_void_p(int(device_mem))
    time_start = time.time()
    dali_tensor.copy_to_external(c_type_pointer)
    print('copy1 | {}'.format(time.time() - time_start))

    #Fast:
    arr = torch.zeros([BATCH_SIZE,3,D_SIZE,D_SIZE],
                      dtype=torch.float32,
                      device=torch.device('cuda', 2))
    c_type_pointer = ctypes.c_void_p(arr.data_ptr())
    time_start = time.time()
    dali_tensor.copy_to_external(c_type_pointer)
    print('copy2 | {}'.format(time.time() - time_start))

**Comment by abcilike** [comment]
ssh://deep_cql@172.18.32.3:22/home/deep_cql/anaconda3/bin/python3.6 -u /home/deep_cql/Project/obs_test/dali_trt/pycuda_test.py
RUN: FileReadPipeline
WARNING: `NormalizePermute` is now deprecated. Use `CropMirrorNormalize` instead
read 1280 files from 1 directories
[<nvidia.dali.backend_impl.TensorListGPU object at 0x7fcd17d69fb8>, <nvidia.dali.backend_impl.TensorListCPU object at 0x7fcd17cfb030>]
copy1 | 0.016083955764770508
copy2 | 0.0010178089141845703
copy1 | 0.0008480548858642578
copy2 | 0.0010569095611572266
copy1 | 0.00084686279296875
copy2 | 0.001180887222290039
copy1 | 0.0008449554443359375
copy2 | 0.00118255615234375
copy1 | 0.0008475780487060547
copy2 | 0.0011830329895019531
copy1 | 0.0008444786071777344
copy2 | 0.0011823177337646484
copy1 | 0.0008490085601806641
copy2 | 0.0011835098266601562
copy1 | 0.0008444786071777344
copy2 | 0.0011839866638183594
copy1 | 0.0008447170257568359
copy2 | 0.0011806488037109375
copy1 | 0.0008497238159179688
copy2 | 0.0011830329895019531
copy1 | 0.0008435249328613281
copy2 | 0.0011785030364990234
copy1 | 0.0008451938629150391
copy2 | 0.0011789798736572266
copy1 | 0.0008444786071777344
copy2 | 0.0011818408966064453
copy1 | 0.0008444786071777344
copy2 | 0.0011823177337646484
copy1 | 0.0008466243743896484
copy2 | 0.0011849403381347656
copy1 | 0.0008447170257568359
copy2 | 0.0011801719665527344
copy1 | 0.0008440017700195312
copy2 | 0.0011799335479736328
copy1 | 0.0008487701416015625
copy2 | 0.0011813640594482422
copy1 | 0.0008442401885986328
copy2 | 0.0011832714080810547
copy1 | 0.0008440017700195312
copy2 | 0.0011832714080810547

Process finished with exit code 0

**Comment by abcilike** [comment]
`
# image dir with plain jpeg files
image_dir = "/home/deep_cql/projects/data/pic_64/test2"
BATCH_SIZE = 64  # batch size per GPU
D_SIZE = 416

import os
os.environ["CUDA_DEVICE"] = "2"
import pycuda.driver as cuda
import pycuda.autoinit
import time
import nvidia.dali.ops as ops
import nvidia.dali.types as types
from nvidia.dali.pipeline import Pipeline
import torch
import ctypes

class CommonPipeline(Pipeline):
  def __init__(self, batch_size, num_threads, device_id):
    super(CommonPipeline, self).__init__(batch_size, num_threads, device_id, exec_async=True,
                                         exec_pipelined=True)

    self.input = ops.ExternalSource()
    self.input_label = ops.ExternalSource()
    self.decode = ops.ImageDecoder(device="mixed", output_type=types.RGB)
    self.resize = ops.Resize(device="gpu",
                             image_type=types.RGB,
                             interp_type=types.INTERP_LINEAR, resize_x=416, resize_y=416)
    self.cmn = ops.NormalizePermute(device="gpu",
                                    output_dtype=types.FLOAT,
                                    height=416,
                                    width=416,
                                    image_type=types.RGB,
                                    mean=[0., 0., 0.],
                                    std=[255., 255., 255.])

  def base_define_graph(self, inputs, labels):
    images = self.decode(inputs)
    # images = self.mix_and_split(images)
    images = self.resize(images)
    output = self.cmn(images)
    return (output, labels)

class FileReadPipeline(CommonPipeline):
  def __init__(self, batch_size, num_threads, device_id):
    super(FileReadPipeline, self).__init__(batch_size, num_threads, device_id)
    self.input = ops.FileReader(file_root=image_dir)
    # self.input = ops.FileReader(file_root=image_dir)

  def define_graph(self):
    images, labels = self.input(name="Reader")
    return self.base_define_graph(images, labels)

if __name__ == '__main__':
  pipe_t = [FileReadPipeline, (0, 1)]
  pipe_name, label_range = pipe_t
  print("RUN: " + pipe_name.__name__)

  pipes = pipe_name(batch_size=BATCH_SIZE, num_threads=3, device_id=2)
  pipes.build()
  pipe_out = pipes.run()
  print(pipe_out)
  dali_tensor, _ = pipe_out

  size = D_SIZE*D_SIZE*3*BATCH_SIZE*4
  device_mem = cuda.mem_alloc(size)
  c_type_pointer1 = ctypes.c_void_p(int(device_mem))

  arr = torch.zeros([BATCH_SIZE, 3, D_SIZE, D_SIZE],
                    dtype=torch.float32,
                    device=torch.device('cuda', 2))
  c_type_pointer2 = ctypes.c_void_p(arr.data_ptr())

  for i in range(20):
    # slow
    time_start = time.time()
    dali_tensor.copy_to_external(c_type_pointer1)
    print('copy1 | {}'.format(time.time() - time_start))

    #Fast:
    time_start = time.time()
    dali_tensor.copy_to_external(c_type_pointer2)
    print('copy2 | {}'.format(time.time() - time_start))`

**Comment by abcilike** [comment]
```
# image dir with plain jpeg files
image_dir = "/home/deep_cql/projects/data/pic_64/test2"
BATCH_SIZE = 64  # batch size per GPU
D_SIZE = 416

import os
os.environ["CUDA_DEVICE"] = "2"
import pycuda.driver as cuda
import pycuda.autoinit
import time
import nvidia.dali.ops as ops
import nvidia.dali.types as types
from nvidia.dali.pipeline import Pipeline
import torch
import ctypes

class CommonPipeline(Pipeline):
  def __init__(self, batch_size, num_threads, device_id):
    super(CommonPipeline, self).__init__(batch_size, num_threads, device_id, exec_async=True,
                                         exec_pipelined=True)

    self.input = ops.ExternalSource()
    self.input_label = ops.ExternalSource()
    self.decode = ops.ImageDecoder(device="mixed", output_type=types.RGB)
    self.resize = ops.Resize(device="gpu",
                             image_type=types.RGB,
                             interp_type=types.INTERP_LINEAR, resize_x=416, resize_y=416)
    self.cmn = ops.NormalizePermute(device="gpu",
                                    output_dtype=types.FLOAT,
                                    height=416,
                                    width=416,
                                    image_type=types.RGB,
                                    mean=[0., 0., 0.],
                                    std=[255., 255., 255.])

  def base_define_graph(self, inputs, labels):
    images = self.decode(inputs)
    # images = self.mix_and_split(images)
    images = self.resize(images)
    output = self.cmn(images)
    return (output, labels)

class FileReadPipeline(CommonPipeline):
  def __init__(self, batch_size, num_threads, device_id):
    super(FileReadPipeline, self).__init__(batch_size, num_threads, device_id)
    self.input = ops.FileReader(file_root=image_dir)
    # self.input = ops.FileReader(file_root=image_dir)

  def define_graph(self):
    images, labels = self.input(name="Reader")
    return self.base_define_graph(images, labels)

if __name__ == '__main__':
  pipe_t = [FileReadPipeline, (0, 1)]
  pipe_name, label_range = pipe_t
  print("RUN: " + pipe_name.__name__)

  pipes = pipe_name(batch_size=BATCH_SIZE, num_threads=3, device_id=2)
  pipes.build()
  pipe_out = pipes.run()
  print(pipe_out)
  dali_tensor, _ = pipe_out

  size = D_SIZE*D_SIZE*3*BATCH_SIZE*4
  device_mem = cuda.mem_alloc(size)
  c_type_pointer1 = ctypes.c_void_p(int(device_mem))

  arr = torch.zeros([BATCH_SIZE, 3, D_SIZE, D_SIZE],
                    dtype=torch.float32,
                    device=torch.device('cuda', 2))
  c_type_pointer2 = ctypes.c_void_p(arr.data_ptr())

  for i in range(20):
    # slow
    time_start = time.time()
    dali_tensor.copy_to_external(c_type_pointer1)
    print('copy1 | {}'.format(time.time() - time_start))

    #Fast:
    time_start = time.time()
    dali_tensor.copy_to_external(c_type_pointer2)
    print('copy2 | {}'.format(time.time() - time_start))
```

**Comment by abcilike** [comment]
```
ssh://deep_cql@172.18.32.3:22/home/deep_cql/anaconda3/bin/python3.6 -u /home/deep_cql/Project/obs_test/dali_trt/pycuda_test.py
RUN: FileReadPipeline
WARNING: `NormalizePermute` is now deprecated. Use `CropMirrorNormalize` instead
read 1280 files from 1 directories
[<nvidia.dali.backend_impl.TensorListGPU object at 0x7f2f49336f10>, <nvidia.dali.backend_impl.TensorListCPU object at 0x7f2f49336f48>]
copy1 | 0.013237953186035156
copy2 | 0.0008099079132080078
copy1 | 0.012892484664916992
copy2 | 0.0007598400115966797
copy1 | 0.012903451919555664
copy2 | 0.0007579326629638672
copy1 | 0.012903690338134766
copy2 | 0.0007562637329101562
copy1 | 0.012869834899902344
copy2 | 0.0007567405700683594
copy1 | 0.012870550155639648
copy2 | 0.0007550716400146484
copy1 | 0.012868881225585938
copy2 | 0.0007562637329101562
copy1 | 0.012887239456176758
copy2 | 0.0007555484771728516
copy1 | 0.012909173965454102
copy2 | 0.0007541179656982422
copy1 | 0.012871742248535156
copy2 | 0.0007569789886474609
copy1 | 0.012853384017944336
copy2 | 0.0007572174072265625
copy1 | 0.012895345687866211
copy2 | 0.0007550716400146484
copy1 | 0.012891530990600586
copy2 | 0.0007538795471191406
copy1 | 0.01288151741027832
copy2 | 0.0007572174072265625
copy1 | 0.012858390808105469
copy2 | 0.0007607936859130859
copy1 | 0.012909650802612305
copy2 | 0.0007557868957519531
copy1 | 0.012866735458374023
copy2 | 0.0007617473602294922
copy1 | 0.012878656387329102
copy2 | 0.0007615089416503906
copy1 | 0.01289677619934082
copy2 | 0.0007617473602294922
copy1 | 0.012874364852905273
copy2 | 0.0007572174072265625

Process finished with exit code 0

```

**Comment by awolant** [comment]
I was able to run your code, but did not reproduce the issue. My times are pretty similar:

```
copy1 | 0.0010552406311035156
copy2 | 0.0007593631744384766
copy1 | 0.000759124755859375
copy2 | 0.0007567405700683594
copy1 | 0.0007638931274414062
copy2 | 0.0007584095001220703
copy1 | 0.0007581710815429688
copy2 | 0.0007541179656982422
copy1 | 0.0007572174072265625
copy2 | 0.00075531005859375
copy1 | 0.0007581710815429688
copy2 | 0.0007543563842773438
copy1 | 0.0007567405700683594
copy2 | 0.0007545948028564453
copy1 | 0.0007560253143310547
copy2 | 0.0007538795471191406
copy1 | 0.0007572174072265625
copy2 | 0.0007538795471191406
copy1 | 0.0007567405700683594
copy2 | 0.0007543563842773438
copy1 | 0.0007586479187011719
copy2 | 0.0007569789886474609
copy1 | 0.0007565021514892578
copy2 | 0.0007548332214355469
copy1 | 0.0007562637329101562
copy2 | 0.0007538795471191406
copy1 | 0.0007579326629638672
copy2 | 0.0007541179656982422
copy1 | 0.0007565021514892578
copy2 | 0.0007555484771728516
copy1 | 0.0007560253143310547
copy2 | 0.0007560253143310547
copy1 | 0.0007560253143310547
copy2 | 0.0007560253143310547
copy1 | 0.0007581710815429688
copy2 | 0.0007565021514892578
copy1 | 0.0007586479187011719
copy2 | 0.0007550716400146484
copy1 | 0.0007576942443847656
copy2 | 0.0007565021514892578
```
What versions of the libraries do you have? What kind of hardware are use using?

**Comment by abcilike** [comment]
ubuntu 16.04
1080TI
cuda_9.0
cudnn 7.3.0
python 3.6
torch 1.0
pycuda 2018.1.1
nvidia-dali  0.12.0

**Comment by abcilike** [comment]
When I change computers, this phenomenon does not exist.  But the previous computer, the problem still exists, I don't know how to solve
```
RUN: FileReadPipeline
read 64 files from 1 directories
[<nvidia.dali.backend_impl.TensorListGPU object at 0x7f3d254d1570>, <nvidia.dali.backend_impl.TensorListCPU object at 0x7f3d254d15a8>]
copy1 | 0.0006878376007080078
copy2 | 0.0005023479461669922
copy1 | 0.0005052089691162109
copy2 | 0.0005002021789550781
copy1 | 0.0005037784576416016
copy2 | 0.000499725341796875
copy1 | 0.0005033016204833984
copy2 | 0.0005004405975341797
copy1 | 0.0005028247833251953
copy2 | 0.000499725341796875
copy1 | 0.0005025863647460938
copy2 | 0.0004992485046386719
copy1 | 0.0005035400390625
copy2 | 0.000499725341796875
copy1 | 0.0005035400390625
copy2 | 0.0005011558532714844
copy1 | 0.0005028247833251953
copy2 | 0.0004992485046386719
copy1 | 0.0005033016204833984
copy2 | 0.0004990100860595703
copy1 | 0.0005028247833251953
copy2 | 0.0004994869232177734
copy1 | 0.0005030632019042969
copy2 | 0.0005004405975341797
copy1 | 0.0005033016204833984
copy2 | 0.0004992485046386719
copy1 | 0.0005025863647460938
copy2 | 0.0004992485046386719
copy1 | 0.0005028247833251953
copy2 | 0.0004992485046386719
copy1 | 0.0005035400390625
copy2 | 0.0005018711090087891
copy1 | 0.0005028247833251953
copy2 | 0.0004994869232177734
copy1 | 0.0005023479461669922
copy2 | 0.0004990100860595703
copy1 | 0.0005028247833251953
copy2 | 0.0004992485046386719
copy1 | 0.0005030632019042969
copy2 | 0.0005013942718505859
```

**Comment by abcilike** [comment]
I created the virtual environment and reinstalled the software library, and the phenomenon still exists.
I think the problem may be caused by cuda version.

**Comment by awolant** [comment]
If that helps, I was using Ubuntu 18.04, CUDA 10.1, PyCUDA 2019.1.2, Torch 1.2.0. 
It would be good, if you could try out some of those to narrow it down a bit.

**Comment by JanuszL** [comment]
Please reopen when you can share a repro.
